operator/internal/manifests/networkpolicy.go

```diff
@@ -0,0 +1,596 @@
+			Namespace: opts.Namespace,
+			Labels:    commonLabels(opts.Name),
+		},
+		Spec: networkingv1.NetworkPolicySpec{
+			PodSelector: *selectorAllLokiComponents,
+			PolicyTypes: []networkingv1.PolicyType{
+				networkingv1.PolicyTypeEgress,
+				networkingv1.PolicyTypeIngress,
+			},
+			Ingress: []networkingv1.NetworkPolicyIngressRule{
+				{
+					From: []networkingv1.NetworkPolicyPeer{lokiComponents(opts.Namespace)},
+					Ports: []networkingv1.NetworkPolicyPort{
+						httpPolicyPort,
+						internalHTTPPolicyPort,
+						grpclbPolicyPort,
+						gossipPolicyPort,
+					},
+				},
+				// Allow ingress for metrics, from Prometheus
+				{
+					From: []networkingv1.NetworkPolicyPeer{
+						promNetworkPolicyPeer,
+					},
+					Ports: []networkingv1.NetworkPolicyPort{
+						httpPolicyPort,
+					},
+				},
+			},
+			Egress: []networkingv1.NetworkPolicyEgressRule{
+				// Egress DNS service
+				egressToDNSK8s,
+				egressToDNSOpenshift,
+				// Egress to common Loki ports for gRPC load balancing & gossip ring
+				{
+					To: []networkingv1.NetworkPolicyPeer{lokiComponents(opts.Namespace)},
+					Ports: []networkingv1.NetworkPolicyPort{
+						httpPolicyPort,
+						internalHTTPPolicyPort,
+						grpclbPolicyPort,
+						gossipPolicyPort,
+					},
+				},
+			},
+		},
+	}
+}
+
+// buildLokiAllowGatewayIngress NetworkPolicy to allow ingress from the
+// gateway to the necessary components.
+func buildLokiAllowGatewayIngress(opts Options) *networkingv1.NetworkPolicy {
+	return &networkingv1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-loki-allow-gateway-ingress", opts.Name),
+			Namespace: opts.Namespace,
+			Labels:    commonLabels(opts.Name),
+		},
+		Spec: networkingv1.NetworkPolicySpec{
+			PodSelector: metav1.LabelSelector{
+				MatchExpressions: []metav1.LabelSelectorRequirement{
+					{
+						Key:      "app.kubernetes.io/name",
+						Operator: metav1.LabelSelectorOpIn,
+						Values:   []string{"lokistack"},
+					},
+					{
+						Key:      "app.kubernetes.io/component",
+						Operator: metav1.LabelSelectorOpIn,
+						Values:   []string{"distributor", "query-frontend", "ruler"},
+					},
+				},
+			},
+			PolicyTypes: []networkingv1.PolicyType{
+				networkingv1.PolicyTypeIngress,
+			},
+			Ingress: []networkingv1.NetworkPolicyIngressRule{
+				{
+					From: []networkingv1.NetworkPolicyPeer{
+						{
+							NamespaceSelector: &metav1.LabelSelector{
+								MatchLabels: map[string]string{
+									"kubernetes.io/metadata.name": opts.Namespace,
+								},
+							},
+							PodSelector: selectorLokiGatewayPods,
+						},
+					},
+					Ports: []networkingv1.NetworkPolicyPort{
+						httpPolicyPort,
+					},
+				},
+			},
+		},
+	}
+}
+
+// buildLokiAllowBucketEgress NetworkPolicy to allow egress traffic from
+// components that need to access object storage to object storage
+func buildLokiAllowBucketEgress(opts Options) *networkingv1.NetworkPolicy {
```

Comment:
Should we try to extract the port from the `endpoint` in the secret? (maybe in addition to defaulting to `443`?)